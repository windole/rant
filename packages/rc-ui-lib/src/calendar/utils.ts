/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
