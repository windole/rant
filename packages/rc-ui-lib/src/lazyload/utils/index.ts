import { ListenEvent } from '../PropsType';

/* eslint-disable prefer-rest-params */
const inBrowser = typeof window !== 'undefined';

export const DEFAULT_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

// 需要监听的事件
export const DEFAULT_EVENTS: ListenEvent[] = [
  'scroll',
  'wheel',
  'resize',
  'mousewheel',
  'animationend',
  'transitionend',
  'touchmove',
];

export const hasIntersectionObserver =
  inBrowser &&
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;

export const modeType = {
  event: 'event',
  observer: 'observer',
};

export const call = (fn): void => fn && fn();

export function setImage(
  element: HTMLMediaElement | HTMLIFrameElement | HTMLImageElement | HTMLElement,
  imagePath: string,
): void {
  if (
    element instanceof HTMLIFrameElement ||
    element instanceof HTMLImageElement ||
    element instanceof HTMLMediaElement
  ) {
    element.src = imagePath;
  } else {
    element.style.backgroundImage = `url('${imagePath}')`;
  }
}

export function loadImage(imagePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = imagePath;
    img.onload = () => resolve(imagePath);
    img.onerror = (err) => reject(err);
  });
}

export function addCssClassName(
  element: HTMLImageElement | HTMLElement,
  cssClassName: string,
): void {
  if (!element.className.includes(cssClassName)) {
    element.className += ` ${cssClassName}`;
  }
}

export function throttle(fn, threshold = 250, scope?) {
  let last;
  let deferTimer;
  return () => {
    // @ts-ignore
    const context = scope || this;

    const now = +new Date();
    const args = arguments;
    if (last && now < last + threshold) {
      // hold on to it
      clearTimeout(deferTimer);
      deferTimer = setTimeout(() => {
        last = now;
        fn.apply(context, args);
      }, threshold);
    } else {
      last = now;
      fn.apply(context, args);
    }
  };
}

export function debounce(func, wait, immediate?) {
  let timeout;
  let args;
  let context;
  let timestamp;
  let result;

  const later = function later() {
    const last = +new Date() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) {
          context = null;
          args = null;
        }
      }
    }
  };

  return function debounced() {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    context = this;
    args = arguments;
    timestamp = +new Date();

    const callNow = immediate && !timeout;
    if (!timeout) {
      timeout = setTimeout(later, wait);
    }

    if (callNow) {
      result = func.apply(context, args);
      context = null;
      args = null;
    }

    return result;
  };
}
