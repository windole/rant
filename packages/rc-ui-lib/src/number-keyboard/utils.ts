import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
