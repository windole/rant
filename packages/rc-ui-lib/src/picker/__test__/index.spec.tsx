/* eslint-disable @typescript-eslint/no-shadow */
import React from 'react';
import { render, cleanup, act, fireEvent } from '@testing-library/react';
import TestsEvent from '../../../tests/events';
import { sleep } from '../../../tests/utils';
import Picker, { PickerInstance, PickerProps } from '..';

const simpleColumn = ['1990', '1991', '1992', '1993', '1994', '1995'];
const columns = [
  {
    values: ['vip', 'normal'],
    className: 'column1',
  },
  {
    values: simpleColumn,
    className: 'column2',
  },
];

describe('Picker', () => {
  function createPicker(props?: Partial<PickerProps>) {
    const pickerRef = React.createRef<PickerInstance>();

    const { queryByTestId, container, rerender, debug } = render(
      <Picker ref={pickerRef} {...props} />,
    );

    return {
      container,
      rerender,
      pickerRef,
      debug,
      queryByTestId,
    };
  }

  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  it('simple columns confirm & cancel event', async () => {
    const onConfirm = jest.fn();
    const onCancel = jest.fn();

    const { container } = createPicker({
      showToolbar: true,
      columns: simpleColumn,
      onConfirm,
      onCancel,
    });

    const confirmBtn = container.querySelector('.rc-picker__confirm');
    const cancelBtn = container.querySelector('.rc-picker__cancel');

    fireEvent.click(confirmBtn);
    fireEvent.click(cancelBtn);

    expect(onConfirm).toHaveBeenCalledWith('1990', 0);
    expect(onCancel).toHaveBeenCalledWith('1990', 0);
  });

  it('multiple columns confirm & cancel event', async () => {
    const onConfirm = jest.fn();
    const onCancel = jest.fn();

    const { container } = createPicker({
      showToolbar: true,
      columns,
      onConfirm,
      onCancel,
    });

    const confirmBtn = container.querySelector('.rc-picker__confirm');
    const cancelBtn = container.querySelector('.rc-picker__cancel');

    fireEvent.click(confirmBtn);
    fireEvent.click(cancelBtn);

    expect(onConfirm).toHaveBeenCalledWith(['vip', '1990'], [0, 0]);
    expect(onCancel).toHaveBeenCalledWith(['vip', '1990'], [0, 0]);
  });

  test('drag columns', async () => {
    const onChange = jest.fn();

    const { container } = createPicker({
      columns,
      onChange,
    });

    const column = container.querySelector('.rc-picker-column__wrapper');
    await TestsEvent.triggerDrag(column, [0, 0]);
    await TestsEvent.triggerDrag(column, [0, -100]);

    await sleep(500);
    expect(onChange).toHaveBeenCalledWith(['normal', '1990'], 0);
  });

  test('drag simple columns', async () => {
    const onChange = jest.fn();

    const { container } = createPicker({
      columns: simpleColumn,
      onChange,
    });

    const column = container.querySelector('.rc-picker-column__wrapper');
    await TestsEvent.triggerDrag(column, [0, -100]);

    await sleep(500);
    expect(onChange).toHaveBeenCalledWith('1995', 5);
  });

  test('column watch default index', async () => {
    const disabled = { disabled: true, text: 1 };
    const props: PickerProps = {
      columns: [disabled, ...simpleColumn],
      itemHeight: 50,
      visibleItemCount: 5,
      swipeDuration: 1000,
    };
    const { container, rerender } = createPicker(props);
    expect(container).toMatchSnapshot();

    const props2: PickerProps = {
      ...props,
      defaultIndex: 2,
    };
    rerender(<Picker {...props2} />);
    expect(container).toMatchSnapshot();
  });

  test('should render title correctly', async () => {
    const { container } = createPicker({
      columns,
      title: 'Custom title',
    });
    expect(container).toMatchSnapshot();
  });

  test('should render toolbar correctly', async () => {
    const { container } = createPicker({
      columns,
      toolbar: 'Custom toolbar',
    });
    expect(container).toMatchSnapshot();
  });

  test('should render option correctly', async () => {
    const { container } = createPicker({
      showToolbar: true,
      columns: ['foo', 'bar'],
      optionRender: (item) => item,
    });
    expect(container).toMatchSnapshot();
  });

  // test('click column item', async () => {
  //   const onChange = jest.fn();
  //   const columns = [
  //     { text: '杭州' },
  //     { text: '宁波' },
  //     { text: '温州', disabled: true },
  //     { text: '嘉兴', disabled: true },
  //   ];
  //   const { container, debug } = createPicker({
  //     columns,
  //     onChange,
  //   });

  //   const columnWrapper = container.querySelector('.rc-picker-column');
  //   await fireEvent.click(columnWrapper);

  //   const column = container.querySelectorAll('.rc-picker-column__item')[2];
  //   debug();
  //   await TestsEvent.triggerDrag(column, [0, 0]);
  //   // await fireEvent.click(column);
  //   expect(onChange).toHaveBeenCalledWith(columns[1], 1);
  // });

  test('toolbar-position prop', () => {
    const { container } = createPicker({
      showToolbar: true,
      toolbarPosition: 'bottom',
      columns,
    });
    expect(container).toMatchSnapshot();
  });

  test('columns-top、columns-bottom prop', () => {
    const { container } = createPicker({
      columnsTop: 'Custom Columns Top',
      columnsBottom: 'Custom Columns Bottom',
      columns,
    });
    expect(container).toMatchSnapshot();
  });

  test('watch columns change', () => {
    const onConfirm = jest.fn();
    const props: PickerProps = {
      showToolbar: true,
      columns: ['1', '2'],
      defaultIndex: 1,
      onConfirm,
    };
    const { container, rerender } = createPicker(props);
    const props2 = {
      ...props,
      columns: ['2', '3'],
    };
    rerender(<Picker {...props2} />);

    fireEvent.click(container.querySelector('.rc-picker__confirm'));
    expect(onConfirm).toHaveBeenCalledWith('3', 1);
  });

  test('should not reset index when columns unchanged', async () => {
    const onConfirm = jest.fn();
    const props: PickerProps = {
      showToolbar: true,
      columns: ['1', '2'],
      onConfirm,
    };
    const { container, rerender, pickerRef } = createPicker(props);
    act(() => {
      pickerRef.current.setIndexes([1]);
    });
    const props2 = {
      ...props,
      columns: ['1', '2'],
    };
    rerender(<Picker {...props2} />);

    fireEvent.click(container.querySelector('.rc-picker__confirm'));
    expect(onConfirm).toHaveBeenCalledWith('2', 1);
  });

  test('set rem item-height', () => {
    const originGetComputedStyle = window.getComputedStyle;

    window.getComputedStyle = () => ({ fontSize: '16px' } as CSSStyleDeclaration);
    const { container } = createPicker({
      columns: simpleColumn.slice(0, 2),
      itemHeight: '10rem',
    });
    expect(container).toMatchSnapshot();

    window.getComputedStyle = originGetComputedStyle;
  });

  test('readonly prop', async () => {
    const onChange = jest.fn();
    const { container } = createPicker({
      columns,
      readonly: true,
      onChange,
    });

    const column = container.querySelectorAll('.rc-picker-column__item')[3];
    await fireEvent.click(column);

    const columnWrapper = container.querySelector('.rc-picker-column__wrapper');
    await TestsEvent.triggerDrag(columnWrapper, [0, -100]);

    expect(onChange).not.toHaveBeenCalled();
  });

  test('set picker values', () => {
    const onConfirm = jest.fn();
    const props: PickerProps = {
      showToolbar: true,
      columns,
      onConfirm,
    };
    const { pickerRef } = createPicker(props);

    const vm = pickerRef.current;

    expect(vm.getColumnValues(-1)).toEqual(undefined);
    expect(vm.getColumnValues(1)).toHaveLength(6);
    expect(vm.getColumnValue(1)).toEqual('1990');

    act(() => {
      vm.setColumnValue(0, 'normal');
    });
    expect(vm.getColumnValue(0)).toEqual('normal');

    act(() => {
      vm.setColumnIndex(0, 0);
    });
    expect(vm.getColumnValue(0)).toEqual('vip');

    act(() => {
      vm.setColumnValue(1, '1991');
    });
    expect(vm.getColumnValue(1)).toEqual('1991');

    act(() => {
      vm.setColumnValues(0, ['vip', 'normal', 'other']);
    });
    expect(vm.getColumnValues(0)).toHaveLength(3);
    expect(vm.getValues()).toHaveLength(2);

    act(() => {
      vm.setColumnValues(-1, []);
    });
    expect(vm.getValues()).toHaveLength(2);

    act(() => {
      vm.setValues(['vip', '1992']);
    });
    expect(vm.getColumnIndex(1)).toEqual(2);
    expect(vm.getColumnIndex(2)).toEqual(0);
    expect(vm.getIndexes()).toEqual([0, 2]);

    act(() => {
      vm.setIndexes([1, 4]);
    });
    expect(vm.getColumnValue(1)).toEqual('1994');
    expect(vm.getColumnValue(2)).toEqual(undefined);
  });
});
