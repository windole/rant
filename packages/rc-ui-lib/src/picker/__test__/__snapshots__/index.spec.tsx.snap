// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker column watch default index 1`] = `
<div>
  <div
    class="rc-picker"
  >
    <div
      class="rc-picker__toolbar"
    >
      <button
        class="rc-picker__cancel"
        type="button"
      >
        取消
      </button>
      <button
        class="rc-picker__confirm"
        type="button"
      >
        确认
      </button>
    </div>
    <div
      class="rc-picker__columns"
      style="height: 250px;"
    >
      <div
        class="rc-picker-column"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,100px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--disabled"
            role="button"
            style="height: 50px;"
            tabindex="-1"
          >
            <div
              class="rc-ellipsis"
            >
              1
            </div>
          </li>
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1990
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1991
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1992
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1993
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1994
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1995
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker__mask"
        style="background-size: 100% 100px;"
      />
      <div
        class="rc-hairline-unset--top-bottom rc-picker__frame"
        style="height: 50px;"
      />
    </div>
  </div>
</div>
`;

exports[`Picker column watch default index 2`] = `
<div>
  <div
    class="rc-picker"
  >
    <div
      class="rc-picker__toolbar"
    >
      <button
        class="rc-picker__cancel"
        type="button"
      >
        取消
      </button>
      <button
        class="rc-picker__confirm"
        type="button"
      >
        确认
      </button>
    </div>
    <div
      class="rc-picker__columns"
      style="height: 250px;"
    >
      <div
        class="rc-picker-column"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,100px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--disabled"
            role="button"
            style="height: 50px;"
            tabindex="-1"
          >
            <div
              class="rc-ellipsis"
            >
              1
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1990
            </div>
          </li>
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1991
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1992
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1993
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1994
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 50px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1995
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker__mask"
        style="background-size: 100% 100px;"
      />
      <div
        class="rc-hairline-unset--top-bottom rc-picker__frame"
        style="height: 50px;"
      />
    </div>
  </div>
</div>
`;

exports[`Picker columns-top、columns-bottom prop 1`] = `
<div>
  <div
    class="rc-picker"
  >
    <div
      class="rc-picker__toolbar"
    >
      <button
        class="rc-picker__cancel"
        type="button"
      >
        取消
      </button>
      <button
        class="rc-picker__confirm"
        type="button"
      >
        确认
      </button>
    </div>
    Custom Columns Top
    <div
      class="rc-picker__columns"
      style="height: 220px;"
    >
      <div
        class="rc-picker-column column1"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              vip
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              normal
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker-column column2"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1990
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1991
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1992
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1993
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1994
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1995
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker__mask"
        style="background-size: 100% 88px;"
      />
      <div
        class="rc-hairline-unset--top-bottom rc-picker__frame"
        style="height: 44px;"
      />
    </div>
    Custom Columns Bottom
  </div>
</div>
`;

exports[`Picker set rem item-height 1`] = `
<div>
  <div
    class="rc-picker"
  >
    <div
      class="rc-picker__toolbar"
    >
      <button
        class="rc-picker__cancel"
        type="button"
      >
        取消
      </button>
      <button
        class="rc-picker__confirm"
        type="button"
      >
        确认
      </button>
    </div>
    <div
      class="rc-picker__columns"
      style="height: 800px;"
    >
      <div
        class="rc-picker-column"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,320px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 160px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1990
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 160px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1991
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker__mask"
        style="background-size: 100% 320px;"
      />
      <div
        class="rc-hairline-unset--top-bottom rc-picker__frame"
        style="height: 160px;"
      />
    </div>
  </div>
</div>
`;

exports[`Picker should render option correctly 1`] = `
<div>
  <div
    class="rc-picker"
  >
    <div
      class="rc-picker__toolbar"
    >
      <button
        class="rc-picker__cancel"
        type="button"
      >
        取消
      </button>
      <button
        class="rc-picker__confirm"
        type="button"
      >
        确认
      </button>
    </div>
    <div
      class="rc-picker__columns"
      style="height: 220px;"
    >
      <div
        class="rc-picker-column"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            foo
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            bar
          </li>
        </ul>
      </div>
      <div
        class="rc-picker__mask"
        style="background-size: 100% 88px;"
      />
      <div
        class="rc-hairline-unset--top-bottom rc-picker__frame"
        style="height: 44px;"
      />
    </div>
  </div>
</div>
`;

exports[`Picker should render title correctly 1`] = `
<div>
  <div
    class="rc-picker"
  >
    <div
      class="rc-picker__toolbar"
    >
      <button
        class="rc-picker__cancel"
        type="button"
      >
        取消
      </button>
      <div
        class="rc-picker__title rc-ellipsis"
      >
        Custom title
      </div>
      <button
        class="rc-picker__confirm"
        type="button"
      >
        确认
      </button>
    </div>
    <div
      class="rc-picker__columns"
      style="height: 220px;"
    >
      <div
        class="rc-picker-column column1"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              vip
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              normal
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker-column column2"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1990
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1991
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1992
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1993
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1994
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1995
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker__mask"
        style="background-size: 100% 88px;"
      />
      <div
        class="rc-hairline-unset--top-bottom rc-picker__frame"
        style="height: 44px;"
      />
    </div>
  </div>
</div>
`;

exports[`Picker should render toolbar correctly 1`] = `
<div>
  <div
    class="rc-picker"
  >
    <div
      class="rc-picker__toolbar"
    >
      Custom toolbar
    </div>
    <div
      class="rc-picker__columns"
      style="height: 220px;"
    >
      <div
        class="rc-picker-column column1"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              vip
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              normal
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker-column column2"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1990
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1991
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1992
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1993
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1994
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1995
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker__mask"
        style="background-size: 100% 88px;"
      />
      <div
        class="rc-hairline-unset--top-bottom rc-picker__frame"
        style="height: 44px;"
      />
    </div>
  </div>
</div>
`;

exports[`Picker toolbar-position prop 1`] = `
<div>
  <div
    class="rc-picker"
  >
    <div
      class="rc-picker__columns"
      style="height: 220px;"
    >
      <div
        class="rc-picker-column column1"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              vip
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              normal
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker-column column2"
      >
        <ul
          aria-hidden="true"
          class="rc-picker-column__wrapper"
          style="transition-duration: 0ms; transition-property: none; transform: translate3d(0,88px,0);"
        >
          <li
            class="rc-picker-column__item rc-picker-column__item--selected"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1990
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1991
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1992
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1993
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1994
            </div>
          </li>
          <li
            class="rc-picker-column__item"
            role="button"
            style="height: 44px;"
            tabindex="0"
          >
            <div
              class="rc-ellipsis"
            >
              1995
            </div>
          </li>
        </ul>
      </div>
      <div
        class="rc-picker__mask"
        style="background-size: 100% 88px;"
      />
      <div
        class="rc-hairline-unset--top-bottom rc-picker__frame"
        style="height: 44px;"
      />
    </div>
    <div
      class="rc-picker__toolbar"
    >
      <button
        class="rc-picker__cancel"
        type="button"
      >
        取消
      </button>
      <button
        class="rc-picker__confirm"
        type="button"
      >
        确认
      </button>
    </div>
  </div>
</div>
`;
